torchtune/modules/transforms/tokenizers/_hf_tokenizer.py

```
@@ -0,0 +1,119 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import json
+from typing import Any, Dict, List, Optional
+
+from tokenizers import Tokenizer
+from torchtune.modules.transforms.tokenizers._utils import BaseTokenizer
+
+
+class HFTokenizer(BaseTokenizer):
+    """
+    A wrapper around Hugging Face tokenizers. This class can be used to load from a
+    Hugging Face tokenizer.json file into a torchtune BaseTokenizer.
+
+    Args:
+        path (str): Path to tokenizer.json file
+        config_path (Optional[str]): Path to tokenizer_config.json file. Default: None
+        generation_config_path (Optional[str]): Path to generation_config.json file.
+            Default: None
+    """
+
+    def __init__(
+        self,
+        path: str,
+        config_path: Optional[str] = None,
+        generation_config_path: Optional[str] = None,
+    ):
+        self.hf_tokenizer = Tokenizer.from_file(path)
+        if config_path:
+            with open(config_path, "rb") as f:
+                self.config = json.load(f)
+        else:
+            self.config = None
+        if generation_config_path:
+            with open(generation_config_path, "rb") as f:
+                self.generation_config = json.load(f)
+        else:
+            self.generation_config = None
+        self._infer_bos_eos_tokens()
```

lets raise an error if self.generation_config is None and self.config is None

Also, apparently we always use one or the other. We could simplify the rest of the code by just making something like:

```
self.config = generation_config or config
```

---

Yeah I think it's simplest if we just use config. At the same time, I read (on some random forum) that generation config should take precedence over config? So currently I'm doing that. One thing that's nice about generation config is that it directly gives the int value for eos and bos, whereas the regular config tends to give the string value, which then needs to be converted. Aside from that, config has chat template, special tokens, and all that other stuff we would need. So yeah right now I am hedging a bit and keeping both around because I suspect we may need both at different times

---

Cite your sources.

---

[Source](https://github.com/huggingface/transformers/issues/25395#issuecomment-1671863257) 

Like I said, maybe not the actual source-of-truth here. Especially as I look through the `PreTrainedTokenizerBase` class -- they seem to really only rely on `tokenizer_config.json` ([ref](https://github.com/huggingface/transformers/blob/main/src/transformers/tokenization_utils_base.py#L2106)). For what it's worth, litgpt also gives `tokenizer_config.json` [precedence](https://github.com/Lightning-AI/litgpt/blob/f6031e3a88e272ec86ad8f412573699589f4d41b/litgpt/tokenizer.py#L30-L40), and it seems to work OK for them (also shout-out to @Andrei-Aksionov who helped give me some pointers here).

So yeah perhaps I will switch the order to read from tokenizer config, then generation config. I do trust the code itself + litgpt's implementation more than that one specific issue comment.

---

Oh, thanks for the shout-out 😊.

It was long time since I worked with tokenizers, but yeah, we decided to rely on `tokenizer.json` and if it's in the repo - use HF Tokenizers. Before `tokenizer.model` and thus SentencePiece had a precedence.

If to take a closer look at the init process of AutoTokenizer, one can see that if there is only a `tokenizer.model` file (a vocab for SP), then this vocab is loaded and then the SP tokenizer is converted to a HF Tokenizers format. The latter allows to extend the vocab with special tokens (often comes with instruct versions of models). With SP to perform the same you will need a much more fiddling (i.e. operate with protobuf format).

So, basically, HF never works with SP tokenizer, even if only it's vocab is provided.
(I [asked](https://huggingface.co/infly/OpenCoder-8B-Instruct/discussions/4) in the repo of OpenCoder what to do if there is only `tokenizer.model`, additional tokens in `tokenizer_config.json` and I don't want to use AutoTokenizer, but never got an answer. Do they all rely on AutoTokenizer?)

In other words, it's good that you are working on supporting `tokenizer.json`.

-----

> For what it's worth, litgpt also gives tokenizer_config.json [precedence](https://github.com/Lightning-AI/litgpt/blob/f6031e3a88e272ec86ad8f412573699589f4d41b/litgpt/tokenizer.py#L30-L40)

FYI, after looking at the codebase, previously `generation_config.json` had a precedence, but it was changed in this  [commit](https://github.com/Lightning-AI/litgpt/pull/552/commits/219662c8672ada6a3b991bb405f9ca89ea79651b).